gluonnlp/model/sampled_block.py

```diff
@@ -0,0 +1,326 @@
+          the input network.
+        - **sampled_candidates**: A tensor of shape `(num_sampled,)`.
+          The sampled candidate classes.
+        - **expected_count_sampled**: A tensor of shape `(num_sampled,)`.
+          The expected count for sampled candidates.
+        - **expected_count_true**: A tensor of shape `(num_sampled)`.
+          The expected count for true classes.
+        - **label**: A tensor of shape `(batch_size,1)`.
+          The target classes.
+
+    Outputs:
+        - **out**: A tensor of shape `(batch_size, 1+num_sampled)`.
+          The output probability for the true class and sampled classes
+        - **new_targets**: A tensor of shape `(batch_size,)`.
+          The new target classes.
+
+    """
+    def __init__(self, num_classes, num_sampled, in_unit, remove_accidental_hits,
+                 prefix=None, params=None):
+        super(SampledLogits, self).__init__(prefix=prefix, params=params)
+        with self.name_scope():
+            self.weight = self.params.get('weight', shape=(num_classes, in_unit),
+                                          grad_stype='row_sparse')
+            self.bias = self.params.get('bias', shape=(num_classes, 1),
+                                        grad_stype='row_sparse')
+        self.block = _SampledLogits(num_classes, num_sampled, in_unit, remove_accidental_hits)
+        self._num_classes = num_classes
+        self._num_sampled = num_sampled
+        self._in_unit = in_unit
+        self._remove_accidental_hits = remove_accidental_hits
+
+    def hybrid_forward(self, F, x, sampled_candidates, expected_count_sampled,
+                       expected_count_true, label, weight, bias):
+        """Forward computation."""
+        # (batch_size,)
+        label = F.reshape(label, shape=(-1,))
+        # (num_sampled+batch_size,)
+        ids = F.concat(sampled_candidates, label, dim=0)
+        # lookup weights and biases
+        # (num_sampled+batch_size, dim)
+        w_all = F.Embedding(data=ids, weight=weight,
+                            input_dim=self._num_classes, output_dim=self._in_unit,
+                            sparse_grad=True)
+        # (num_sampled+batch_size, 1)
+        b_all = F.Embedding(data=ids, weight=bias,
+                            input_dim=self._num_classes, output_dim=1,
+                            sparse_grad=True)
+        return self.block(x, sampled_candidates, expected_count_sampled,
+                          expected_count_true, label, w_all, b_all)
+
+    def __repr__(self):
+        s = '{name}({mapping})'
+        mapping = '{0} -> {1}, with {2} samples'.format(self._in_unit, self._num_classes,
+                                                        self._num_sampled)
+        return s.format(name=self.__class__.__name__,
+                        mapping=mapping,
+                        **self.__dict__)
+
+class SparseSampledLogits(Block):
```

Comment:
Renamed to SparseISorNCELogitsNet, and move to model.train. The computation logics are shared with the Dense one, so it would be great in the future we can merge together.